chained/water_impl.go

```diff
@@ -0,0 +1,57 @@
+package chained
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"log/slog"
+	"net"
+
+	"github.com/getlantern/common/config"
+	"github.com/getlantern/flashlight/v7/ops"
+	"github.com/refraction-networking/water"
+	_ "github.com/refraction-networking/water/transport/v1"
+)
+
+type waterImpl struct {
+	raddr          string
+	reportDialCore reportDialCoreFn
+	dialer         water.Dialer
+	nopCloser
+}
+
+func newWaterImpl(addr string, pc *config.ProxyConfig, reportDialCore reportDialCoreFn) (*waterImpl, error) {
+	b64WASM := ptSetting(pc, "water_wasm")
+	wasm, err := base64.StdEncoding.DecodeString(b64WASM)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode water wasm: %w", err)
+	}
+
+	cfg := &water.Config{
+		TransportModuleBin: wasm,
+		OverrideLogger:     slog.New(newLogHandler(log, ptSetting(pc, "water_transport"))),
+	}
+
+	dialer, err := water.NewDialerWithContext(context.Background(), cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create dialer: %w", err)
+	}
+
+	return &waterImpl{
+		raddr:          addr,
+		dialer:         dialer,
+		reportDialCore: reportDialCore,
+	}, nil
+}
+
+func (d *waterImpl) dialServer(op *ops.Op, ctx context.Context) (net.Conn, error) {
+	return d.reportDialCore(op, func() (net.Conn, error) {
+		conn, err := d.dialer.DialContext(context.Background(), "tcp", d.raddr)
```

Comment:
We do actually want to respect the dial context.  The issue is a bug in `water`.  I pinged the WATER team on Slack about it [here](https://wdynhnkxvsdx.slack.com/archives/C06GJBFP8C8/p1721341403567499?thread_ts=1720463188.453499&cid=C06GJBFP8C8).  Essentially the library hanging onto the dial context beyond its scope (just the dial).  If you cancel this context, all connections dialed with the context will be closed.  This should not happen - only dials in progress should be affected.

Here's a little example program to demonstrate, with a comparison to a simple TCP dialer/listener pair from the net package:

https://github.com/hwh33/water-context-bug

We can just mark this as a TODO for now though.  Could you add a TODO comment here explaining the bug and the reason for ignoring the dial context?  Then we can merge this PR and continue to make progress with integration.  When the WATER team fixes the issue, we can circle back and use `ctx` for the dial here.

Reply:
Comment looks great!